File: tests/node/test_workflow.mjs
/**
 * E2E tests for the Blazen Node.js bindings.
 *
 * Uses the ava test runner.
 * Build the native binding first:
 *   cd crates/blazen-node && npm install && npm run build
 */

import test from "ava";

import { Workflow, Context, version } from "../../crates/blazen-node/index.js";

// =========================================================================
// Version
// =========================================================================

test("version · returns a string", (t) => {
  const v = version();
  t.is(typeof v, "string");
  t.truthy(v.length > 0);
});

// =========================================================================
// Single step echo
// =========================================================================

test("single step echo · echoes the input back as the result", async (t) => {
  const wf = new Workflow("echo");
  wf.addStep("echo", ["blazen::StartEvent"], async (event, ctx) => {
    return { type: "blazen::StopEvent", result: { message: event.message } };
  });

  const result = await wf.run({ message: "hello e2e" });
  t.is(result.type, "blazen::StopEvent");
  t.is(result.data.message, "hello e2e");
});

// =========================================================================
// Multi-step pipeline
// =========================================================================

test("multi-step pipeline · chains two steps via custom event type", async (t) => {
  const wf = new Workflow("pipeline");

  wf.addStep("analyze", ["blazen::StartEvent"], async (event, ctx) => {
    return {
      type: "AnalyzeEvent",
      text: event.message,
      length: event.message.length,
    };
  });

  wf.addStep("finalize", ["AnalyzeEvent"], async (event, ctx) => {
    return {
      type: "blazen::StopEvent",
      result: { text: event.text, length: event.length },
    };
  });

  const result = await wf.run({ message: "hello world" });
  t.is(result.data.text, "hello world");
  t.is(result.data.length, 11);
});

// =========================================================================
// Context sharing
// =========================================================================

test("context sharing · shares state between steps via context", async (t) => {
  const wf = new Workflow("ctx-test");

  wf.addStep("setter", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("counter", 42);
    return { type: "NextEvent" };
  });

  wf.addStep("getter", ["NextEvent"], async (event, ctx) => {
    const val = await ctx.get("counter");
    return { type: "blazen::StopEvent", result: { counter: val } };
  });

  const result = await wf.run({});
  t.is(result.data.counter, 42);
});

// =========================================================================
// Context run ID
// =========================================================================

test("context runId · returns a UUID string", async (t) => {
  const wf = new Workflow("run-id");

  wf.addStep("capture", ["blazen::StartEvent"], async (event, ctx) => {
    const rid = await ctx.runId();
    return { type: "blazen::StopEvent", result: { run_id: rid } };
  });

  const result = await wf.run({});
  const rid = result.data.run_id;
  t.is(typeof rid, "string");
  t.truthy(rid.length > 0);
  // UUID format: 8-4-4-4-12
  t.is(rid.split("-").length, 5);
});

// =========================================================================
// Streaming
// =========================================================================

test("streaming · receives intermediate events via runStreaming", async (t) => {
  const wf = new Workflow("stream-test");

  wf.addStep("producer", ["blazen::StartEvent"], async (event, ctx) => {
    for (let i = 0; i < 3; i++) {
      await ctx.writeEventToStream({ type: "Progress", step: i });
    }
    return { type: "blazen::StopEvent", result: { done: true } };
  });

  const collected = [];
  const result = await wf.runStreaming({}, (event) => {
    collected.push(event);
  });

  t.is(result.data.done, true);

  const progress = collected.filter((e) => e.type === "Progress");
  t.truthy(
    progress.length > 0,
    `expected at least one Progress event, got ${progress.length}`
  );
});

// =========================================================================
// Step returns array (fan-out)
// =========================================================================

test("fan-out · handles step returning an array of events", async (t) => {
  const wf = new Workflow("fan-out");

  wf.addStep("fan", ["blazen::StartEvent"], async (event, ctx) => {
    return [
      { type: "BranchA", value: "a" },
      { type: "BranchB", value: "b" },
    ];
  });

  wf.addStep("handle_a", ["BranchA"], async (event, ctx) => {
    return { type: "blazen::StopEvent", result: { branch: "a" } };
  });

  wf.addStep("handle_b", ["BranchB"], async (event, ctx) => {
    return { type: "blazen::StopEvent", result: { branch: "b" } };
  });

  const result = await wf.run({});
  t.truthy(
    result.data.branch === "a" || result.data.branch === "b",
    `unexpected branch: ${result.data.branch}`
  );
});

// =========================================================================
// Step returns null (side-effect + sendEvent continuation)
// =========================================================================

test("step returns null · continues via ctx.sendEvent when step returns null", async (t) => {
  const wf = new Workflow("null-test");

  wf.addStep("side_effect", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("processed", true);
    await ctx.sendEvent({ type: "Continue" });
    return null;
  });

  wf.addStep("finisher", ["Continue"], async (event, ctx) => {
    const processed = await ctx.get("processed");
    return { type: "blazen::StopEvent", result: { processed } };
  });

  const result = await wf.run({});
  t.is(result.data.processed, true);
});

// =========================================================================
// setBytes / getBytes roundtrip
// =========================================================================

test("setBytes / getBytes roundtrip · stores and retrieves binary data as Buffer", async (t) => {
  const wf = new Workflow("bytes-roundtrip");

  wf.addStep("bytes", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.setBytes("bin", Buffer.from([0xde, 0xad, 0xbe, 0xef]));
    const buf = await ctx.getBytes("bin");
    return {
      type: "blazen::StopEvent",
      result: { bytes: Array.from(buf) },
    };
  });

  const result = await wf.run({});
  t.deepEqual(result.data.bytes, [0xde, 0xad, 0xbe, 0xef]);
});

// =========================================================================
// get() on bytes key returns array
// =========================================================================

test("get() on bytes key returns array · returns a JSON array of numbers instead of null", async (t) => {
  const wf = new Workflow("bytes-get");

  wf.addStep("bytes", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.setBytes("bin", Buffer.from([0xde, 0xad, 0xbe, 0xef]));
    const val = await ctx.get("bin");
    return {
      type: "blazen::StopEvent",
      result: { val },
    };
  });

  const result = await wf.run({});
  t.not(result.data.val, null);
  t.deepEqual(result.data.val, [222, 173, 190, 239]);
});

// =========================================================================
// Complex JSON roundtrip
// =========================================================================

test("complex JSON roundtrip · stores and retrieves nested objects, arrays, and mixed types", async (t) => {
  const wf = new Workflow("complex-json");
  const payload = {
    nested: { arr: [1, 2.5, null, true, "str"], obj: { a: 1 } },
  };

  wf.addStep("complex", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("complex", payload);
    const val = await ctx.get("complex");
    return { type: "blazen::StopEvent", result: { val } };
  });

  const result = await wf.run({});
  t.deepEqual(result.data.val, payload);
});

// =========================================================================
// Overwrite behavior
// =========================================================================

test("overwrite behavior · overwrites a key with a value of a different type", async (t) => {
  const wf = new Workflow("overwrite");

  wf.addStep("overwrite", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("k", 1);
    await ctx.set("k", "hello");
    const val = await ctx.get("k");
    return { type: "blazen::StopEvent", result: { val } };
  });

  const result = await wf.run({});
  t.is(result.data.val, "hello");
});

// =========================================================================
// Null / undefined handling
// =========================================================================

test("null / undefined handling · stores null and returns null for missing keys", async (t) => {
  const wf = new Workflow("null-handling");

  wf.addStep("nulls", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("k", null);
    const stored = await ctx.get("k");
    const missing = await ctx.get("no_such_key");
    return {
      type: "blazen::StopEvent",
      result: { stored, missing },
    };
  });

  const result = await wf.run({});
  t.is(result.data.stored, null);
  t.is(result.data.missing, null);
});

test("null / undefined handling · returns default when key is missing or null", async (t) => {
  const wf = new Workflow("get-default");

  wf.addStep("defaults", ["blazen::StartEvent"], async (event, ctx) => {
    await ctx.set("present", 5);
    await ctx.set("nullish", null);
    const buf = Buffer.from([1, 2]);
    return {
      type: "blazen::StopEvent",
      result: {
        missingDefault: await ctx.get("missing", "fb"),
        missingNumber: await ctx.get("missing", 0),
        presentIgnoresDefault: await ctx.get("present", 99),
        nullishUsesDefault: await ctx.get("nullish", "fb"),
        missingNoDefault: await ctx.get("missing"),
        missingBytesDefault: Array.from(
          await ctx.getBytes("missing", buf),
        ),
        stateMissingDefault: await ctx.state.get("missing", "sd"),
        sessionMissingDefault: await ctx.session.get("missing", "ssd"),
      },
    };
  });

  const result = await wf.run({});
  t.is(result.data.missingDefault, "fb");
  t.is(result.data.missingNumber, 0);
  t.is(result.data.presentIgnoresDefault, 5);
  t.is(result.data.nullishUsesDefault, "fb");
  t.is(result.data.missingNoDefault, null);
  t.deepEqual(result.data.missingBytesDefault, [1, 2]);
  t.is(result.data.stateMissingDefault, "sd");
  t.is(result.data.sessionMissingDefault, "ssd");
});

// =========================================================================
// Timeout
// =========================================================================

test("timeout · times out when step takes too long", async (t) => {
  const wf = new Workflow("timeout-test");
  wf.setTimeout(0.1); // 100ms

  wf.addStep("slow", ["blazen::StartEvent"], async (event, ctx) => {
    await new Promise((r) => setTimeout(r, 10000));
    return { type: "blazen::StopEvent", result: {} };
  });

  await t.throwsAsync(() => wf.run({}), {
    message: /timeout|timed out/i,
  });
});

File: tests/python/test_workflow.py
"""E2E tests for the Blazen Python bindings.

These tests exercise the full workflow lifecycle through the PyO3 bindings:
event creation, step decoration, workflow execution, context sharing,
streaming, and fan-out.
"""

import asyncio

import pytest

from blazen import (
    Context,
    Event,
    StartEvent,
    StopEvent,
    Workflow,
    step,
)


# =========================================================================
# Event creation
# =========================================================================


def test_event_creation():
    ev = Event("AnalyzeEvent", text="hello", score=0.9)
    assert ev.event_type == "AnalyzeEvent"
    assert ev.text == "hello"
    assert ev.score == 0.9


def test_event_to_dict():
    ev = Event("TestEvent", x=1, y="two")
    d = ev.to_dict()
    assert d["x"] == 1
    assert d["y"] == "two"


def test_start_event():
    ev = StartEvent(message="hi")
    assert ev.event_type == "blazen::StartEvent"
    assert ev.message == "hi"


def test_stop_event():
    ev = StopEvent(result={"answer": 42})
    assert ev.event_type == "blazen::StopEvent"
    assert ev.result == {"answer": 42}


def test_event_attribute_error():
    ev = Event("Foo", x=1)
    with pytest.raises(AttributeError):
        _ = ev.nonexistent_field


# =========================================================================
# Single step echo (async)
# =========================================================================


@pytest.mark.asyncio
async def test_single_step_echo():
    @step
    async def echo(ctx: Context, ev: Event):
        return StopEvent(result=ev.to_dict())

    wf = Workflow("echo", [echo])
    handler = await wf.run(message="hello")
    result = await handler.result()

    assert result.event_type == "blazen::StopEvent"
    assert result.result["message"] == "hello"


# =========================================================================
# Single step echo (sync)
# =========================================================================


@pytest.mark.asyncio
async def test_single_step_echo_sync():
    @step
    def echo(ctx: Context, ev: Event):
        return StopEvent(result=ev.to_dict())

    wf = Workflow("echo-sync", [echo])
    handler = await wf.run(message="hello")
    result = await handler.result()

    assert result.event_type == "blazen::StopEvent"
    assert result.result["message"] == "hello"


# =========================================================================
# Multi-step pipeline
# =========================================================================


@pytest.mark.asyncio
async def test_multi_step_pipeline():
    @step
    async def analyze(ctx: Context, ev: Event):
        return Event("AnalyzeDone", text=ev.message, length=len(ev.message))

    @step(accepts=["AnalyzeDone"])
    async def finalize(ctx: Context, ev: Event):
        return StopEvent(result={"text": ev.text, "length": ev.length})

    wf = Workflow("pipeline", [analyze, finalize])
    handler = await wf.run(message="hello world")
    result = await handler.result()

    assert result.result["text"] == "hello world"
    assert result.result["length"] == 11


# =========================================================================
# Context set / get
# =========================================================================


@pytest.mark.asyncio
async def test_context_set_get():
    @step
    async def setter(ctx: Context, ev: Event):
        ctx.set("counter", 42)
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    async def getter(ctx: Context, ev: Event):
        val = ctx.get("counter")
        return StopEvent(result={"counter": val})

    wf = Workflow("ctx-test", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["counter"] == 42


# =========================================================================
# Context run_id
# =========================================================================


@pytest.mark.asyncio
async def test_context_run_id():
    run_id_holder = {}

    @step
    async def capture_id(ctx: Context, ev: Event):
        rid = ctx.run_id()
        run_id_holder["id"] = rid
        return StopEvent(result={"run_id": rid})

    wf = Workflow("run-id-test", [capture_id])
    handler = await wf.run()
    result = await handler.result()

    rid = result.result["run_id"]
    assert isinstance(rid, str)
    assert len(rid) > 0
    # UUID format: 8-4-4-4-12
    assert rid.count("-") == 4


# =========================================================================
# Streaming
# =========================================================================


@pytest.mark.asyncio
async def test_streaming():
    @step
    async def producer(ctx: Context, ev: Event):
        for i in range(3):
            ctx.write_event_to_stream(
                Event("Progress", step=i)
            )
        return StopEvent(result={"done": True})

    wf = Workflow("stream-test", [producer])
    handler = await wf.run()

    collected = []
    async for event in handler.stream_events():
        collected.append(event)

    result = await handler.result()
    assert result.result["done"] is True

    # We should have received at least some progress events.
    progress = [e for e in collected if e.event_type == "Progress"]
    assert len(progress) > 0


# =========================================================================
# Step returns list (fan-out)
# =========================================================================


@pytest.mark.asyncio
async def test_step_returns_list():
    @step
    async def fan_out(ctx: Context, ev: Event):
        return [
            Event("BranchA", value="a"),
            Event("BranchB", value="b"),
        ]

    @step(accepts=["BranchA"])
    async def handle_a(ctx: Context, ev: Event):
        return StopEvent(result={"branch": "a"})

    @step(accepts=["BranchB"])
    async def handle_b(ctx: Context, ev: Event):
        return StopEvent(result={"branch": "b"})

    wf = Workflow("fan-out", [fan_out, handle_a, handle_b])
    handler = await wf.run()
    result = await handler.result()

    # One of the two branches wins.
    assert result.result["branch"] in ("a", "b")


# =========================================================================
# Step returns None (side-effect + ctx.send_event continuation)
# =========================================================================


@pytest.mark.asyncio
async def test_step_returns_none():
    @step
    async def side_effect(ctx: Context, ev: Event):
        ctx.set("processed", True)
        ctx.send_event(Event("Continue"))
        return None

    @step(accepts=["Continue"])
    async def finisher(ctx: Context, ev: Event):
        processed = ctx.get("processed")
        return StopEvent(result={"processed": processed})

    wf = Workflow("none-test", [side_effect, finisher])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["processed"] is True


# =========================================================================
# Context bytes roundtrip
# =========================================================================


@pytest.mark.asyncio
async def test_context_bytes_roundtrip():
    @step
    def setter(ctx: Context, ev: Event):
        ctx.set("k", b"\x00\x01\x02")
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        val = ctx.get("k")
        assert isinstance(val, bytes)
        return StopEvent(result={"data": list(val)})

    wf = Workflow("bytes-roundtrip", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["data"] == [0, 1, 2]


# =========================================================================
# Context complex JSON roundtrip
# =========================================================================


@pytest.mark.asyncio
async def test_context_complex_json():
    @step
    def setter(ctx: Context, ev: Event):
        ctx.set("k", {"nested": [1, 2.5, None, True, "str"]})
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        val = ctx.get("k")
        return StopEvent(result=val)

    wf = Workflow("complex-json", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result == {"nested": [1, 2.5, None, True, "str"]}


# =========================================================================
# Context pickle roundtrip
# =========================================================================


@pytest.mark.asyncio
async def test_context_pickle_roundtrip():
    from dataclasses import dataclass

    @dataclass
    class Payload:
        name: str
        score: float

    @step
    def setter(ctx: Context, ev: Event):
        ctx.set("k", Payload(name="test", score=3.14))
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        val = ctx.get("k")
        return StopEvent(result={"name": val.name, "score": val.score})

    wf = Workflow("pickle-roundtrip", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["name"] == "test"
    assert abs(result.result["score"] - 3.14) < 1e-9


# =========================================================================
# Context set_bytes / get_bytes
# =========================================================================


@pytest.mark.asyncio
async def test_context_set_bytes_get_bytes():
    @step
    def setter(ctx: Context, ev: Event):
        ctx.set_bytes("k", b"\xff\xfe")
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        val = ctx.get_bytes("k")
        return StopEvent(result={"data": list(val)})

    wf = Workflow("set-bytes-get-bytes", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["data"] == [0xFF, 0xFE]


# =========================================================================
# Context overwrite
# =========================================================================


@pytest.mark.asyncio
async def test_context_overwrite():
    @step
    def setter(ctx: Context, ev: Event):
        ctx.set("k", 1)
        ctx.set("k", "hello")
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        val = ctx.get("k")
        return StopEvent(result={"value": val})

    wf = Workflow("overwrite", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["value"] == "hello"


# =========================================================================
# Context buffer not mangled (bytearray vs array.array)
# =========================================================================


@pytest.mark.asyncio
async def test_context_buffer_not_mangled():
    import array

    @step
    def setter(ctx: Context, ev: Event):
        # bytearray should be stored as raw bytes (tier 1)
        ctx.set("ba", bytearray(b"\xaa\xbb"))
        # array.array is NOT bytes/bytearray, so it should pickle (tier 3)
        ctx.set("arr", array.array("i", [1, 2, 3]))
        return Event("NextEvent")

    @step(accepts=["NextEvent"])
    def getter(ctx: Context, ev: Event):
        ba = ctx.get("ba")
        arr = ctx.get("arr")
        return StopEvent(
            result={
                "ba_is_bytes": isinstance(ba, bytes),
                "ba_data": list(ba),
                "arr_type": type(arr).__name__,
                "arr_data": arr.tolist(),
            }
        )

    wf = Workflow("buffer-not-mangled", [setter, getter])
    handler = await wf.run()
    result = await handler.result()

    assert result.result["ba_is_bytes"] is True
    assert result.result["ba_data"] == [0xAA, 0xBB]
    assert result.result["arr_type"] == "array"
    assert result.result["arr_data"] == [1, 2, 3]
